sphinx/theming.py

```diff
@@ -11,212 +11,143 @@
-
-        tdir, tinfo = self.themes[name]
-        if tinfo is None:
+        if path.isdir(theme_path):
             # already a directory, do nothing
             self.rootdir = None
-            self.themedir = tdir
-            self.themedir_created = False
+            self.themedir = theme_path
         else:
             # extract the theme to a temp directory
             self.rootdir = tempfile.mkdtemp('sxt')
             self.themedir = path.join(self.rootdir, name)
-            self.themedir_created = True
-            ensuredir(self.themedir)
-            for name in tinfo.namelist():
-                if name.endswith('/'):
-                    continue
-                dirname = path.dirname(name)
-                if not path.isdir(path.join(self.themedir, dirname)):
-                    os.makedirs(path.join(self.themedir, dirname))
-                with open(path.join(self.themedir, name), 'wb') as fp:
-                    fp.write(tinfo.read(name))
-
-        self.themeconf = configparser.RawConfigParser()
-        self.themeconf.read(path.join(self.themedir, THEMECONF))  # type: ignore
+            extract_zip(theme_path, self.themedir)
+
+        self.config = configparser.RawConfigParser()
+        self.config.read(path.join(self.themedir, THEMECONF))  # type: ignore
 
         try:
-            inherit = self.themeconf.get('theme', 'inherit')
+            inherit = self.config.get('theme', 'inherit')
         except configparser.NoOptionError:
-            raise ThemeError('theme %r doesn\'t have "inherit" setting' % name)
+            raise ThemeError(_('theme %r doesn\'t have "inherit" setting') % name)
 
-        # load inherited theme automatically #1794, #1884, #1885
-        self.load_extra_theme(inherit)
+        if inherit != 'none':
+            try:
+                self.base = factory.create(inherit)
+            except ThemeError:
+                raise ThemeError(_('no theme named %r found, inherited by %r') %
+                                 (inherit, name))
 
-        if inherit == 'none':
-            self.base = None
-        elif inherit not in self.themes:
-            raise ThemeError('no theme named %r found, inherited by %r' %
-                             (inherit, name))
+    @property
+    def dirs(self):
+        # type: () -> List[unicode]
+        """Return a list of theme directories, beginning with this theme's,
+        then the base theme's, then that one's base theme's, etc.
+        """
+        if self.base is None:
+            return [self.themedir]
         else:
-            self.base = Theme(inherit)
+            return [self.themedir] + self.base.get_dirchain()
 
-    def get_confstr(self, section, name, default=NODEFAULT):
+    def get_dirchain(self):
+        # type: () -> List[unicode]
+        """Return a list of theme directories, beginning with this theme's,
+        then the base theme's, then that one's base theme's, etc.
+        """
+        return self.dirs
```

Comment:
I think this is a backward-compatibility method. However, this Theme class is used only from Sphinx inside. IMO, this compat method is not needed.